plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +151,91 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkRequestSource implements Iterator<ActionRequest> {
+
+        final BulkRequest bulkRequest;
+        final ActionListener<BulkResponse> actionListener;
+
+        final Set<Integer> failedSlots;
+        final List<BulkItemResponse> itemResponses;
+
+        int currentSlot = -1;
+
+        BulkRequestSource(BulkRequest bulkRequest, ActionListener<BulkResponse> actionListener) {
+            this.bulkRequest = bulkRequest;
+            this.actionListener = actionListener;
+            this.failedSlots = new HashSet<>();
+            this.itemResponses = new ArrayList<>(bulkRequest.requests().size());
+        }
+
+        @Override
+        public ActionRequest next() {
+            return bulkRequest.requests().get(++currentSlot);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (currentSlot + 1) < bulkRequest.requests().size();
+        }
+
+        Tuple<BulkRequest, ActionListener<BulkResponse>> bulkRequestAndActionListener() {
```

Comment:
returning a Tuple here doesn't look great to me. I would love if we could simplify this. Can we maybe move this logic out of this class and do something along the same lines in `processBulkIndexRequest` instead? Ideally this class would simply hold the items and iterate over them, the creation of the new bulk request and the new action listener could happen outside of it, I think that should make things cleaner.